File: src/ios/Founditure/Models/Listing.swift
// Foundation framework - Latest
import Foundation
// CoreLocation framework - Latest
import CoreLocation

/// Human Tasks:
/// 1. Configure proper data encryption for sensitive listing information
/// 2. Set up monitoring for AI recognition service availability
/// 3. Verify location permission handling in the app
/// 4. Review listing expiration policy with business team
/// 5. Configure proper logging for listing status changes

// MARK: - Enums

/// Represents the condition of the furniture item
/// Requirements addressed:
/// - Data Types (1.3): Furniture listings data structure
public enum FurnitureCondition: String, Codable {
    case excellent
    case good
    case fair
    case poor
}

/// Categories of furniture items
/// Requirements addressed:
/// - AI-powered furniture recognition (1.2): Furniture categorization
public enum FurnitureCategory: String, Codable {
    case chair
    case table
    case sofa
    case bed
    case storage
    case other
}

/// Current status of the listing
/// Requirements addressed:
/// - Data Types (1.3): Furniture listings lifecycle management
public enum ListingStatus: String, Codable {
    case active
    case pending
    case collected
    case expired
}

// MARK: - Listing Class

/// Represents a furniture item listing with comprehensive details and location information
/// Requirements addressed:
/// - AI-powered furniture recognition (1.2): AI-based furniture categorization
/// - Location-based discovery (1.2): Location tracking for furniture items
/// - Data Types (1.3): Comprehensive furniture listing data model
public final class Listing: Codable, Identifiable {
    // MARK: - Properties
    
    public let id: UUID
    public let title: String
    public let description: String
    public let category: FurnitureCategory
    public let condition: FurnitureCondition
    public let location: Location
    public let userId: UUID
    public let imageUrls: [String]
    public let aiTags: [String]
    public let aiConfidenceScore: Double
    public private(set) var status: ListingStatus
    public let createdAt: Date
    public private(set) var collectedAt: Date?
    public private(set) var expiresAt: Date?
    
    // MARK: - Initialization
    
    /// Initializes a new Listing instance with the provided details
    public init(
        title: String,
        description: String,
        category: FurnitureCategory,
        condition: FurnitureCondition,
        location: Location,
        userId: UUID,
        imageUrls: [String],
        aiTags: [String],
        aiConfidenceScore: Double
    ) {
        self.id = UUID()
        self.title = title
        self.description = description
        self.category = category
        self.condition = condition
        self.location = location
        self.userId = userId
        self.imageUrls = imageUrls
        self.aiTags = aiTags
        self.aiConfidenceScore = aiConfidenceScore
        self.status = .active
        self.createdAt = Date()
        self.expiresAt = Calendar.current.date(byAdding: .day, value: 7, to: self.createdAt)
        self.collectedAt = nil
    }
    
    // MARK: - Public Methods
    
    /// Updates the listing status and related timestamps
    /// Requirements addressed:
    /// - Data Types (1.3): Listing lifecycle management
    public func updateStatus(_ newStatus: ListingStatus) {
        self.status = newStatus
        
        switch newStatus {
        case .collected:
            self.collectedAt = Date()
        case .expired:
            self.expiresAt = Date()
        default:
            break
        }
    }
    
    /// Checks if the listing has expired
    /// Requirements addressed:
    /// - Data Types (1.3): Listing lifecycle management
    public func isExpired() -> Bool {
        guard let expiresAt = expiresAt else {
            return false
        }
        return Date() > expiresAt
    }
    
    /// Calculates distance from another location
    /// Requirements addressed:
    /// - Location-based discovery (1.2): Distance calculation for furniture discovery
    public func distanceFrom(_ otherLocation: Location) -> Double {
        return location.distanceTo(otherLocation)
    }
    
    // MARK: - Codable Implementation
    
    private enum CodingKeys: String, CodingKey {
        case id, title, description, category, condition, location
        case userId, imageUrls, aiTags, aiConfidenceScore
        case status, createdAt, collectedAt, expiresAt
    }
    
    public required init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        id = try container.decode(UUID.self, forKey: .id)
        title = try container.decode(String.self, forKey: .title)
        description = try container.decode(String.self, forKey: .description)
        category = try container.decode(FurnitureCategory.self, forKey: .category)
        condition = try container.decode(FurnitureCondition.self, forKey: .condition)
        location = try container.decode(Location.self, forKey: .location)
        userId = try container.decode(UUID.self, forKey: .userId)
        imageUrls = try container.decode([String].self, forKey: .imageUrls)
        aiTags = try container.decode([String].self, forKey: .aiTags)
        aiConfidenceScore = try container.decode(Double.self, forKey: .aiConfidenceScore)
        status = try container.decode(ListingStatus.self, forKey: .status)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        collectedAt = try container.decodeIfPresent(Date.self, forKey: .collectedAt)
        expiresAt = try container.decodeIfPresent(Date.self, forKey: .expiresAt)
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        
        try container.encode(id, forKey: .id)
        try container.encode(title, forKey: .title)
        try container.encode(description, forKey: .description)
        try container.encode(category, forKey: .category)
        try container.encode(condition, forKey: .condition)
        try container.encode(location, forKey: .location)
        try container.encode(userId, forKey: .userId)
        try container.encode(imageUrls, forKey: .imageUrls)
        try container.encode(aiTags, forKey: .aiTags)
        try container.encode(aiConfidenceScore, forKey: .aiConfidenceScore)
        try container.encode(status, forKey: .status)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encodeIfPresent(collectedAt, forKey: .collectedAt)
        try container.encodeIfPresent(expiresAt, forKey: .expiresAt)
    }
}

File: src/android/app/src/main/java/com/founditure/domain/model/Listing.kt
package com.founditure.domain.model

import android.os.Parcelable // Latest Android SDK version
import kotlinx.parcelize.Parcelize // kotlinx.parcelize:1.9.0
import com.founditure.domain.model.User.id

/**
 * Human Tasks:
 * 1. Ensure Android SDK is properly configured in the project's build.gradle
 * 2. Verify kotlinx.parcelize plugin is enabled in the module's build.gradle
 * 3. Configure ProGuard/R8 rules if using code obfuscation to preserve Parcelable implementation
 * 4. Set up proper listing expiration threshold in configuration
 */

/**
 * Enum class representing possible states of a furniture listing
 * 
 * Requirements addressed:
 * - Core Features - Furniture listings (1.3 Scope/Core Features): 
 *   Defines standardized listing states for furniture item lifecycle management
 */
enum class ListingStatus {
    AVAILABLE,
    PENDING,
    COLLECTED,
    EXPIRED,
    REMOVED
}

/**
 * Enum class representing the condition of the furniture item
 * 
 * Requirements addressed:
 * - Core Features - Furniture listings (1.3 Scope/Core Features): 
 *   Standardizes furniture condition classification for consistent item quality assessment
 */
enum class FurnitureCondition {
    EXCELLENT,
    GOOD,
    FAIR,
    POOR
}

/**
 * Data class representing a furniture listing with all its details.
 * 
 * Requirements addressed:
 * - Core Features - Furniture listings (1.3 Scope/Core Features): 
 *   Implements comprehensive furniture listing data structure
 * - AI-powered furniture recognition (1.3 Scope/Core Features):
 *   Supports AI-generated tags for furniture categorization
 * - Location-based discovery (1.3 Scope/Core Features):
 *   Includes geolocation data for furniture items
 */
@Parcelize
data class Listing(
    /**
     * Unique identifier for the listing
     */
    val id: String,
    
    /**
     * Reference to the user who created the listing
     */
    val userId: String,
    
    /**
     * Title/name of the furniture item
     */
    val title: String,
    
    /**
     * Detailed description of the furniture item
     */
    val description: String,
    
    /**
     * Current status of the listing
     */
    val status: ListingStatus,
    
    /**
     * Physical condition of the furniture
     */
    val condition: FurnitureCondition,
    
    /**
     * List of URLs to the furniture item images
     */
    val imageUrls: List<String>,
    
    /**
     * Geographical latitude of the furniture location
     */
    val latitude: Double,
    
    /**
     * Geographical longitude of the furniture location
     */
    val longitude: Double,
    
    /**
     * Human-readable address of the furniture location
     */
    val address: String,
    
    /**
     * AI-generated tags and categories for the furniture
     */
    val aiTags: Map<String, String>,
    
    /**
     * Timestamp when the listing was created
     */
    val postedAt: Long,
    
    /**
     * Timestamp of the last update to the listing
     */
    val updatedAt: Long
) : Parcelable {

    companion object {
        // Listing expiration threshold in milliseconds (30 days)
        private const val EXPIRATION_THRESHOLD = 30L * 24 * 60 * 60 * 1000
    }

    /**
     * Checks if the listing is still available for collection
     *
     * @return true if the listing status is AVAILABLE
     */
    fun isAvailable(): Boolean {
        return status == ListingStatus.AVAILABLE
    }

    /**
     * Checks if the listing has expired based on posting time
     *
     * @return true if the listing has exceeded the expiration threshold
     */
    fun isExpired(): Boolean {
        val currentTime = System.currentTimeMillis()
        return (currentTime - postedAt) > EXPIRATION_THRESHOLD
    }

    /**
     * Converts the domain model to a database entity
     *
     * @return ListingEntity instance created from this domain model
     */
    fun toListingEntity(): ListingEntity {
        return ListingEntity(
            id = id,
            userId = userId,
            title = title,
            description = description,
            status = status.name,
            condition = condition.name,
            imageUrls = imageUrls.joinToString(","),
            latitude = latitude,
            longitude = longitude,
            address = address,
            aiTags = aiTags.entries.joinToString(",") { "${it.key}:${it.value}" },
            postedAt = postedAt,
            updatedAt = updatedAt
        )
    }
}

/**
 * Data class representing the database entity for a Listing
 * Note: This is a simplified version - actual implementation should match your database schema
 */
data class ListingEntity(
    val id: String,
    val userId: String,
    val title: String,
    val description: String,
    val status: String,
    val condition: String,
    val imageUrls: String,
    val latitude: Double,
    val longitude: Double,
    val address: String,
    val aiTags: String,
    val postedAt: Long,
    val updatedAt: Long
)